cpu/stm32/periph/usbdev_fs.c

```diff
@@ -0,0 +1,661 @@
+#include "cpu_conf.h"
+#include "periph/pm.h"
+#include "periph/gpio.h"
+#include "periph/usbdev.h"
+#include "usbdev_stm32.h"
+#include <string.h>
+/**
+ * Be careful with enabling debug here. As with all timing critical systems it
+ * is able to interfere with USB functionality and you might see different
+ * errors than debug disabled
+ */
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+#ifndef USB1_BASE
+#define USB1_BASE    USB_BASE
+#endif
+
+#ifndef USB1_PMAADDR
+#define USB1_PMAADDR USB_PMAADDR
+#endif /* ndef USB1_PMAADDR */
+
+#define _ENDPOINT_NUMOF     (8) /* IP specific */
+#define STM32_FS_BUF_SPACE (1024)
+/* Base Buffer address located in USB SRAM area */
+#define BUF_BASE_OFFSET   (_ENDPOINT_NUMOF * 8)
+/* List of instantiated USB peripherals */
+static stm32_usbdev_fs_t _usbdevs[USBDEV_NUMOF];
+
+static usbdev_ep_t _ep_in[_ENDPOINT_NUMOF];
+static usbdev_ep_t _ep_out[_ENDPOINT_NUMOF];
+
+static uint8_t* _ep_out_buf[_ENDPOINT_NUMOF];
+static uint8_t* _app_pbuf[_ENDPOINT_NUMOF];
+static uint32_t _ep_in_buf[_ENDPOINT_NUMOF];
+
+/* Forward declaration for the usb device driver */
+const usbdev_driver_t driver;
+
+static USB_TypeDef *_global_regs(const stm32_usbdev_fs_config_t *conf)
+{
+    return (USB_TypeDef*)conf->base_addr;
+}
+
+/* Endpoint Buffer Descriptor */
+typedef struct {
+    uint16_t addr_tx;
+    uint16_t count_tx;
+    uint16_t addr_rx;
+    uint16_t count_rx;
+} ep_buf_desc_t;
+
+#define EP_DESC ((ep_buf_desc_t*)USB1_PMAADDR)
+#define EP_REG(x) (*((volatile uint32_t*) (USB1_BASE + (4*x))))
+
+usbdev_t *usbdev_get_ctx(unsigned num)
+{
+    assert(num < USBDEV_NUMOF);
+    return &_usbdevs[num].usbdev;
+}
+
+static inline void _enable_irq(void)
+{
+    stm32_usbdev_fs_t *usbdev = &_usbdevs[0];
+    const stm32_usbdev_fs_config_t *conf = usbdev->config;
+    _global_regs(conf)->CNTR = USB_CNTR_CTRM | USB_CNTR_RESETM  |
+                               USB_CNTR_ERRM | USB_CNTR_PMAOVRM;
+}
+
+static void _setup(stm32_usbdev_fs_t *usbdev,
+                      const stm32_usbdev_fs_config_t *config)
+{
+    usbdev->usbdev.driver = &driver;
+    usbdev->config = config;
+    usbdev->out = _ep_out;
+    usbdev->in = _ep_in;
+}
+
+static uint32_t _type_to_reg(usb_ep_type_t type)
+{
+    switch (type) {
+        case USB_EP_TYPE_CONTROL:
+            return USB_EP_CONTROL;
+        case USB_EP_TYPE_ISOCHRONOUS:
+            return USB_EP_ISOCHRONOUS;
+        case USB_EP_TYPE_BULK:
+            return USB_EP_BULK;
+        case USB_EP_TYPE_INTERRUPT:
+            return USB_EP_INTERRUPT;
+        default:
+            assert(false);
+            return 0;
+    }
+}
+
+void usbdev_init_lowlevel(void)
+{
+    for (size_t i = 0; i < USBDEV_NUMOF; i++) {
+        _setup(&_usbdevs[i], &stm32_usbdev_fs_config[i]);
+    }
+}
+
+/*TODO: rework this function (maybe move it outside this file)
+        for other ST families support */
+static void _enable_usb_clk(void)
+{
+    /* Validate USB Supply */
+    PWR->CR2 |= PWR_CR2_USV;
```

Comment:
```suggestion
#if defined(PWR_CR2_USV)
    /* Validate USB Supply */
    PWR->CR2 |= PWR_CR2_USV;
#endif
```
Since this define is not available on all families that this driver might support (I'm trying on l0).

Reply:
Added.